source/docs/software/advanced-gradlerio/profiling-with-visualvm.rst

```diff
@@ -0,0 +1,176 @@
+
+.. image:: images/visualvm/visualvm-function-profiling.png
+   :alt: Analyzing the VisualVM function timing tree
+   :width: 700
+
+The above screenshot shows a breakdown of the total time a specific function takes. You can see that ``totallyNotSlowFunction()`` accounts for ``61.9%`` of the robot program CPU time. We can then correlate this to our robot program. In ``totallyNotSlowFunction()``, we have the following code.
+
+.. code-block:: Java
+
+   public static void totallyNotSlowFunction() {
+      for (int i = 0; i < 2000; i++) {
+         System.out.println("HAHAHAHA");
+      }
+   }
+
+In this code snippet, we can identify 2 major causes of concern. A long running ``for`` loop blocks the rest of the robot program from running. Additionally, ``System.out.println()`` calls on the roboRIO are typically quite expensive. We found this information by profiling the Java application on the roboRIO!
+
+Creating a Heap Dump
+--------------------
+
+Besides viewing the remote systems CPU and memory usage, VisualVM is most useful by creating a **Heap Dump**. When a Java object is created, it resides in an area of memory called the heap. When the heap is full, a process called `garbage collection <https://www.geeksforgeeks.org/garbage-collection-java/>`__ begins. Garbage collection can be a common cause of loop overruns in a traditional Java robot program.
+
+To begin, ensure you are on the :guilabel:`Monitor` tab and click :guilabel:`Heap Dump`.
+
+.. image:: images/visualvm/visualvm-perform-heapdump.png
+   :alt: Location of heap dump button in VisualVM
+   :width: 700
+
+This heap dump will be stored on the target system (roboRIO) and must be retrieved using SFTP. For this tutorial, `FileZilla <https://filezilla-project.org/>`__ will be the tool of choice. Open FileZilla and connect to the roboRIO with the following parameters
+
+- Host: ``sftp://172.22.11.2`` or ``sftp://10.TE.AM.2``
+- Username: ``lvuser``
+- Password: Leave blank
+- Port: Leave blank
+
+If connected successfully, you'll see a list of files on the roboRIO.
+
+.. image:: images/visualvm/filezilla-connection.png
+   :alt: Filezilla list of files
+   :width: 700
+
+Navigate to the ``/tmp/`` directory with FileZilla and download it to your machine. Once downloaded, the dump can be analyzed with VisualVM.
+
+Analyzing a Heap Dump
+---------------------
+
+Reopen VisualVM if closed using the previous instructions. Then click on :guilabel:`File` and :guilabel:`Load`. Navigate to the retrieved dump file and load it.
+
+.. image:: images/visualvm/visualvm-viewing-dump.png
+   :alt: Viewing a dump in VisualVM
+   :width: 700
+
+Clicking on :guilabel:`Summary` and selecting :guilabel:`Objects` instead will show a breakdown of objects by quantity. The below screenshot showcases a completely empty robot program, and then one that creates an million large ``ArrayList`` of integers.
```

Comment:
Is there a way to see where those million integers is created? I'm not sure there's much useful for teams based on what's shown in the screenshot.

Reply:
Hm, I'll have to find that out myself

Reply:
In dotMemory, it shows the creation call stack, I wonder if VisualVM has something similar.